stl/error_handler.py

```diff
@@ -17,43 +17,85 @@
-                                 ' "pre_states" must be non-empty.')
-
-_ERROR_ID_EMPTY_PRE_STATES = 204
-_ERROR_MSG_EMPTY_PRE_STATES = ('Transitions require non-empty "pre_states".')
-
-# Lex erros live in the 300s
-_ERROR_ID_UNKNOWN_LEX_ERROR = 300
-_ERROR_MSG_UNKNOWN_LEX_ERROR = 'There was an unknown lexing/tokenizing error :('
-
-_ERROR_ID_MISSING_STRING_QUOTE = 301
-_ERROR_MSG_MISSING_STRING_QUOTE = 'Missing " " on string literal.'
-
-_ERROR_ID_SINGLE_QUOTE = 302
-_ERROR_MSG_SINGLE_QUOTE = 'Use double quotes (" ") for string literals.'
+import stl.lexer_error
+import stl.parser_error
+
+
+# To generate new parser errors:
+#   1. Create a new STL file that contains the new error
+#   2. Run `stl/parser.py your_error.stl`
+#   3. The parser will dumb out the symbol stack in the logs
+#   4. Take the symbols from the end of the stack that are representative
+#          of your error to create a "stack pattern".
+#   5. Create a ParserError object, and add it to the list of errors in
+#          ParserErrorHandler.GetError
+
+# Low level parse errors live in the 200s
+_UNKNOWN_PARSER_ERROR = stl.parser_error.ParserError(
+    error_name='unknown-parser-error',
+    error_id=200,
+    error_msg='There was an unknown parse error :(',
+    stack_patterns=None)
+_MISSING_SEMICOLON = stl.parser_error.ParserError(
+    error_name='missing-semicolon',
+    error_id=201,
+    error_msg='Missing semicolon.',
+    stack_patterns=[
+        ['CONST', 'type', 'NAME'],
+        ['MODULE', 'NAME'],
+        ['TRANSITION', 'NAME', 'params', '=', 'NAME', '(', 'param_values_without_paren', ')'],
+        ['NAME', 'ARROW', 'NAME', 'param_values', 'ARROW', 'NAME'],
+    ])
+_MISSING_CLOSING_CURLY_BRACE = stl.parser_error.ParserError(
+    error_name='missing-closing-curly-brace',
+    error_id=202,
+    error_msg='Missing closing curly brace }.',
+    stack_patterns=[
+        ['{'],
+    ])
+
+# Higer level parse errors live in the 300s
+_MISSING_POST_STATES = stl.parser_error.ParserError(
+    error_name='missing-post-states',
+    error_id=301,
+    error_msg=('Transitions require "post_states". If there'
+               ' are no explicit post_states use an empty'
+               ' list (post_states = []).'),
+    stack_patterns=[
+        ['TRANSITION', 'NAME', 'params', '{', 'local_vars', 'pre_states', 'events'],
+    ])
+_MISSING_PRE_STATES = stl.parser_error.ParserError(
+    error_name='missing-pre-states',
+    error_id=302,
+    error_msg=('Transitions require "pre_states", and the'
+               ' "pre_states" must be non-empty.'),
+    stack_patterns=[
+        ['TRANSITION', 'NAME', 'params', '{'],
+    ])
+_EMPTY_PRE_STATES = stl.parser_error.ParserError(
+    error_name='empty-pre-states',
+    error_id=303,
+    error_msg='Transitions require non-empty "pre_states".',
+    stack_patterns=[
+        ['TRANSITION', 'NAME', 'params', '{', 'local_vars', 'PRE_STATES', '=', '[']
+    ])
+
+# Lexer errors live in the 100s
```

Comment:
You may want to put lexer errors first since their error_id is smaller than parsers.

Reply:
Done